### PR TITLE
Implement AMMs router

### LINKED ISSUE #457: Implement AMMs Router
**Motivation**

Enable to route orders between different `AMM` instances or pools. We chose to not make it part of the `AMM` pallet due to the fact that the `AMM` pallet can be instantiated many times, and we may want to route orders between all of them (for example we could have a route `USDT-KSM-xKSM`).

### PR DESCRIPTION
**Pull request by rustaceanstayhigh**
Closed #457

### CODE REVIEW

File: pallets/router/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    )>;
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+        type PalletId: Get<PalletId>;
+
+        /// Specify all the AMMs we are routing between
+        type AMMs: Get<Route<Self>>;
```
Comment (ETeissonniere) [inline pallets/router/src/lib.rs:63]: This should be a vec of instances implementing the AMM trait

File: pallets/router/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+        /// Trade interface
+        type AMM: AMM<Self::AccountId, CurrencyId, BalanceOf<Self>>;
```
Comment (ETeissonniere) [inline pallets/router/src/lib.rs:66]: This should be parts of `AMMs`

File: pallets/router/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+            #[pallet::compact] expiry: BlockNumberFor<T>,
+        ) -> DispatchResultWithPostInfo {
+            let trader = ensure_signed(origin)?;
+
+            // Ensure the length of routers should be >= 1 at least.
+            ensure!(!route.is_empty(), Error::<T>::EnptyRouters);
+
+            // Ensure balances user input is bigger than zero.
+            ensure!(
+                amount_in > Zero::zero() && min_amount_out > Zero::zero(),
```
Comment (ETeissonniere) [inline pallets/router/src/lib.rs:114]: I guess min amount could be 0

File: pallets/router/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    }
+
+    #[pallet::pallet]
+    #[pallet::generate_store(pub(super) trait Store)]
+    pub struct Pallet<T>(_);
+
+    #[pallet::error]
+    pub enum Error<T> {
+        BalanceLow,
+        EnptyRouters,
```
Comment (ashi20192020ayush) [inline pallets/router/src/lib.rs:78]:
```suggestion
        EmptyRouters,
```

Spelling Mistake

File: pallets/router/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        type Currency: MultiCurrencyExtended<Self::AccountId>;
+    }
+
+    #[pallet::pallet]
+    #[pallet::generate_store(pub(super) trait Store)]
+    pub struct Pallet<T>(_);
+
+    #[pallet::error]
+    pub enum Error<T> {
+        BalanceLow,
```
Comment (ashi20192020ayush) [inline pallets/router/src/lib.rs:77]: Please add comment before each errors like other pallets

File: pallets/router/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+        /// [sender, amount_in, route, amount_out]
+        TradedSuccessfully(T::AccountId, BalanceOf<T>, Route<T>, BalanceOf<T>),
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        #[pallet::weight(10_000)]
```
Comment (ashi20192020ayush) [inline pallets/router/src/lib.rs:98]: Can we add comment here?

**Review by ashi20192020ayush — DISMISSED** [review]

File: pallets/router/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # Route for Automatic Market Maker (AMM)
+//!
+//! Given a supported `route` like this `(0, USDT, KSM)`, we can get corresponding AMM or pool.
```
Comment (ETeissonniere) [inline pallets/router/src/lib.rs:17]:
```suggestion
//! # Router for Automatic Market Maker (AMM)
//!
//! Given a supported `route`, executes the indicated trades on all the available AMM(s) pool(s).
```

File: pallets/router/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+
+            // Ensure the routes user input are valid.
+            ensure!(
+                route.iter().all(|r| all_routes.iter().any(|_r| r == _r)),
+                Error::<T>::NotSupportedRoute
+            );
+
+            let mut amount_out: Balance = Zero::zero();
+            for sub_route in route.iter() {
+                let (_id, from_currency_id, to_currency_id) = sub_route;
```
Comment (ETeissonniere) [inline pallets/router/src/lib.rs:175]:
```suggestion
                let (from_currency_id, to_currency_id) = sub_route;
```

File: pallets/router/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+
+    pub type Route = Vec<(
+        // ID of the AMM to use, as specified in the `Config` trait. Setting this
+        // to 0 would take the first AMM instance specified in `type AMMs`.
+        u8,
+        // Base asset
+        CurrencyId,
+        // Quote asset
+        CurrencyId,
+    )>;
```
Comment (ETeissonniere) [inline pallets/router/src/lib.rs:52]:
```suggestion
    pub type Route = Vec<(
        // Base asset
        CurrencyId,
        // Quote asset
        CurrencyId,
    )>;
```

for now we only have one AMM due to our simplification efforts

File: pallets/router/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+        /// Route pallet id
+        #[pallet::constant]
+        type RoutePalletId: Get<PalletId>;
```
Comment (ETeissonniere) [inline pallets/router/src/lib.rs:60]:
```suggestion
        /// Router pallet id
        #[pallet::constant]
        type RouterPalletId: Get<PalletId>;
```

File: pallets/router/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+
+        /// Specify all the AMMs we are routing between
+        type AMMAdaptor: AMM<Self>;
```
Comment (ETeissonniere) [inline pallets/router/src/lib.rs:63]:
```suggestion
        type AMM: AMM<Self>;
```

only one AMM and no adapter in this simplified design

File: pallets/router/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+
+        /// Specify all the AMMs we are routing between
+        type Routes: Get<Route>;
```
Comment (ETeissonniere) [inline pallets/router/src/lib.rs:66]:
```suggestion
```

File: pallets/router/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+
+        /// How many routes we support at most
+        #[pallet::constant]
+        type MaxLengthRoute: Get<u8>;
```
Comment (ETeissonniere) [inline pallets/router/src/lib.rs:70]: we could modify `Route` from `Vec` to a `BoundedVec` to avoid this

File: pallets/router/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+            for sub_route in route.iter() {
+                let (_id, from_currency_id, to_currency_id) = sub_route;
+
+                amount_out = amount_in;
+                amount_out = T::AMMAdaptor::trade(
+                    &trader,
+                    (*from_currency_id, *to_currency_id),
+                    amount_out,
+                    min_amount_out,
+                )?;
```
Comment (ETeissonniere) [inline pallets/router/src/lib.rs:183]:
```suggestion
                amount_out = T::AMMAdaptor::trade(
                    &trader,
                    (*from_currency_id, *to_currency_id),
                    amount_in,
                    min_amount_out,
                )?;
```

this got me confused 😉

File: pallets/router/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+
+            // Get all AMM routes we're supporting now.
+            let all_routes = T::Routes::get();
+
+            // Ensure the routes user input are valid.
+            ensure!(
+                route.iter().all(|r| all_routes.iter().any(|_r| r == _r)),
+                Error::<T>::NotSupportedRoute
+            );
+
```
Comment (ETeissonniere) [inline pallets/router/src/lib.rs:172]:
```suggestion
```

I believe this is unnecessary

File: pallets/router/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+        InsufficientBalance,
+        /// Input wrong route that we don't support now
+        NotSupportedRoute,
+        /// The expiry is smaller than current block number
+        TooSmallExpiry,
+        /// Exceed the max length of routes we allow
+        ExceedMaxLengthRoute,
+        /// Input duplicated route
+        DuplicatedRoute,
+    }
```
Comment (ETeissonniere) [inline pallets/router/src/lib.rs:100]:
```suggestion
    #[pallet::error]
    pub enum Error<T> {
        /// Input balance must not be zero
        ZeroBalance,
        /// Must input one route at least
        EmptyRoute,
        /// User hasn't enough tokens for transaction
        InsufficientBalance,
        /// The expiry is smaller than current block number
        TooSmallExpiry,
        /// Exceed the max length of routes we allow
        ExceedMaxLengthRoute,
        /// Input duplicated route
        DuplicatedRoute,
    }
```

File: pallets/router/src/mock.rs [hunk 17]
```diff
@@ -0,0 +1,234 @@
...
+
+pub(crate) fn run_to_block(n: u64) {
+    while System::block_number() < n {
+        AMMRoute::on_finalize(System::block_number());
+        System::on_finalize(System::block_number());
+        System::set_block_number(System::block_number() + 1);
+        System::on_initialize(System::block_number());
+        AMMRoute::on_initialize(System::block_number());
+    }
+}
```
Comment (ETeissonniere) [inline pallets/router/src/mock.rs:234]: we don't rely on the `on_initialize` and `on_finalize` functions for this pallet. Thus we should be able to replace this as below:
```suggestion
pub(crate) fn run_to_block(n: u64) {
    System::set_block_number(n);
}
```

File: pallets/router/src/tests.rs [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+        // pools values should be updated - we should have less XDOT
+        assert_eq!(DOT2XDOT::pools(XDOT, DOT).unwrap().base_amount, 99_999_006);
+
+        // pools values should be updated - we should have more DOT in the pool
+        assert_eq!(
+            DOT2XDOT::pools(XDOT, DOT).unwrap().quote_amount,
+            100_000_998
+        );
+    })
+}
```
Comment (ETeissonniere) [inline pallets/router/src/tests.rs:119]: This test is great however it doesn't cover the case of a cross pool trade (where we have more than one route). Let's add one to cover this.

**Review by ETeissonniere — CHANGES_REQUESTED** [review]

File: pallets/router/src/lib.rs [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+                amount_out = T::AMM::trade(
+                    &trader,
+                    (*from_currency_id, *to_currency_id),
+                    amount_in,
+                    min_amount_out,
+                )?;
+
+                amount_in = amount_out;
+            }
+
```
Comment (ETeissonniere) [inline pallets/router/src/lib.rs:173]:
```suggestion
            for sub_route in route.iter() {
                let (from_currency_id, to_currency_id) = sub_route;

                amount_out = T::AMM::trade(
                    &trader,
                    (*from_currency_id, *to_currency_id),
                    amount_in,
                    1,
                )?;

                amount_in = amount_out;
            }
            
            ensure!(amount_out >= min_amount_out, Error::<T>::UnexpectedSlippage);

```

File: pallets/router/src/lib.rs [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+        /// Must input one route at least
+        EmptyRoute,
+        /// User hasn't enough tokens for transaction
+        InsufficientBalance,
+        /// The expiry is smaller than current block number
+        TooSmallExpiry,
+        /// Exceed the max length of routes we allow
+        ExceedMaxLengthRoute,
+        /// Input duplicated route
+        DuplicatedRoute,
```
Comment (ETeissonniere) [inline pallets/router/src/lib.rs:95]:
```suggestion
        /// Input duplicated route
        DuplicatedRoute,
        /// We received less coins than the minimum amount specified
        UnexpectedSlippage,
```

**Review by ETeissonniere — APPROVED** [review]

**Comment by drbh** [comment]
@rustaceanstayhigh LGTM 👌 except for the CI build error 

```
the package `parallel-runtime` depends on `pallet-router`, with features: `pallet-balances` but `pallet-router` does not have these features.
```

**Pull request merged**